gitlab: 8.0.5 -> 8.5.1, service improvements

fpletz opened the pull request:
###### Things done:
- [x] Tested via `nix.useChroot`.
- [x] Built on platform(s): x86_64
- [x] Tested compilation of all pkgs that depend on this change: none
- [x] Tested execution of binary products.
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @roblabla @cstrahan

---

Updates gitlab to the current stable version and fixes a lot of features that
were broken, at least with the current version and our configuration.

Quite a lot of sweat and tears has gone into testing nearly all features and
reading/patching the Gitlab source as we're about to deploy gitlab for our
whole company.

Things to note:
- The gitlab config is now written as a nix attribute set and will be
  converted to JSON. Gitlab uses YAML but JSON is a subset of YAML.
  The `extraConfig` opition is also an attribute set that will be merged
  with the default config. This way _all_ Gitlab options are supported.
- Some paths like uploads and configs are hardcoded in rails  (at least
  after my study of the Gitlab source). This is why they are linked from
  the Gitlab root to /run/gitlab and then linked to the  configurable
  `statePath`.
- Backup & restore should work out of the box from another Gitlab instance.
- gitlab-git-http-server has been replaced by gitlab-workhorse upstream.
  Push & pull over HTTPS works perfectly. Communication to gitlab is done
  over unix sockets. An HTTP server is required to proxy requests to
  gitlab-workhorse over another unix socket at
  `/run/gitlab/gitlab-workhorse.socket`.
- The user & group running gitlab are now configurable. These can even be
  changed for live instances.
- The initial email address & password of the root user can be configured.
###### TODO
- [x] Documentation + Release Notes (fixes #13303)
- [x] Gitlab 8.5.1 was released

fpletz committed:
gitlab: 8.0.5 -> 8.5.0, service improvements

Updates gitlab to the current stable version and fixes a lot of features that
were broken, at least with the current version and our configuration.

Quite a lot of sweat and tears has gone into testing nearly all features and
reading/patching the Gitlab source as we're about to deploy gitlab for our
whole company.

Things to note:

 * The gitlab config is now written as a nix attribute set and will be
   converted to JSON. Gitlab uses YAML but JSON is a subset of YAML.
   The `extraConfig` opition is also an attribute set that will be merged
   with the default config. This way *all* Gitlab options are supported.

 * Some paths like uploads and configs are hardcoded in rails  (at least
   after my study of the Gitlab source). This is why they are linked from
   the Gitlab root to /run/gitlab and then linked to the  configurable
   `statePath`.

 * Backup & restore should work out of the box from another Gitlab instance.

 * gitlab-git-http-server has been replaced by gitlab-workhorse upstream.
   Push & pull over HTTPS works perfectly. Communication to gitlab is done
   over unix sockets. An HTTP server is required to proxy requests to
   gitlab-workhorse over another unix socket at
   `/run/gitlab/gitlab-workhorse.socket`.

 * The user & group running gitlab are now configurable. These can even be
   changed for live instances.

 * The initial email address & password of the root user can be configured.

Fixes #8598.

jagajaga commented:
LGTM

fpletz commented:
I will add documention and an entry to the release notes before merging this. Feedback on the code would be much appreciated.

zimbatm commented on nixos/modules/services/misc/gitlab.nix:
```diff
@@ -21,14 +21,15 @@ let
-    user: gitlab
-    gitlab_url: "http://${cfg.host}:${toString cfg.port}/"
+    user: ${cfg.user}
+    gitlab_url: "http://localhost:8080/"
```
Is this true, http://localhost:8080 is a constant ?

fpletz replied:
Yup, if you're using this nixos module this is constant because the listening socket is hardcoded in the unicorn config. The problem here was that gitlab-shell somehow wouldn't work with https (I changed the scheme of course :panda_face:).

But thinking about it, we should probably make this configurable somehow because 8080 is a very heavily used default port that could be already in use by some other service. I will check if gitlab shell supports unix sockets, then we can get rid of the tcp socket altogether. I will look into it and make it at least configurable.

zimbatm replied:
Awesome

zimbatm commented on nixos/modules/services/misc/gitlab.nix:
```diff
@@ -37,33 +38,101 @@ let
   gitlab-runner = pkgs.stdenv.mkDerivation rec {
     name = "gitlab-runner";
-    buildInputs = [ pkgs.gitlab pkgs.bundler pkgs.makeWrapper ];
+    buildInputs = with pkgs; [ gitlab bundler makeWrapper ];
```
the gitlab package could be made an option so that users can switch versions more easily, and then have gitlab.bundler as a passthrough.

fpletz replied:
Sounds useful. I'm not sure however how tightly coupled the module will be to specific gitlab versions in the future. But it makes the gitlab package itself override-able which could be beneficial for deploying patch versions if we're not catching up in time after the release, so I'll add that. Thanks.

fpletz replied:
This should also be done for gitlab-shell and gitlab-workhorse. Also, these components could even be be put into the gitlab store path because every gitlab release mandates specific versions of these components.

zimbatm commented on pkgs/applications/version-management/gitlab/remove-hardcoded-locations.patch:
```diff
@@ -18,74 +18,10 @@ index 3316ece..c34dec0 100644
- 
-     # Uncomment and customize if you can't use the default user to run GitLab (default: 'git')
--    # user: git
-+    user: gitlab
```
is this constant correct ?

fpletz replied:
This part of the patch was removed. :wink:

zimbatm replied:
Cool :sunglasses:

zimbatm commented:
I'm done with reviewing the changes. It's really nice work.

fpletz committed:
fixup! gitlab: 8.0.5 -> 8.5.0, service improvements

Some debugging and development leftovers.

cstrahan commented:
@fpletz Awesome work - thanks! LGTM :)

zimbatm commented:
@fpletz ping me here or on IRC when you're good to go

fpletz committed:
fixup! gitlab: 8.0.5 -> 8.5.0, service improvements

gitlab-shell is now talking over the unix socket to gitlab, so the TCP
port isn't needed anymore.

fpletz committed:
fixup! gitlab: 8.0.5 -> 8.5.0, service improvements

8.5.0 -> 8.5.1

fpletz committed:
gitlab: Add documentation

Fixes #13303.

fpletz committed:
fixup! gitlab: 8.0.5 -> 8.5.0, service improvements

Make the gitlab packages and components overrideable.

fpletz commented:
@zimbatm Ready for the final review. I've addressed the issues we discussed above & added documentation. Also, 8.5.1 was released in the meantime.

If you give your thumbs up on the changes, I would like to proceed to squash the commits, do a final reword of the main commit and push to master.

domenkozar commented:
Would be nice to check if this renders https://github.com/NixOS/nixpkgs/pull/8599 obsolete
